clinica/utils/inputs.py

```diff
@@ -251,14 +260,40 @@ def clinica_file_reader(
             that do not exists in the existing file path.
 
     """
+    from multiprocessing import Manager
     from os.path import join
 
+    from joblib import Parallel, delayed
+
     from clinica.utils.exceptions import (
         ClinicaBIDSError,
         ClinicaCAPSError,
         ClinicaException,
     )
 
+    manager = Manager()
+
+    def find_sub_ses_pattern(input_directory, sub, ses, error_encountered, results):
+        if is_bids:
+            origin_pattern = join(input_directory, sub, ses)
+        else:
+            origin_pattern = join(input_directory, "subjects", sub, ses)
+
+        current_pattern = join(origin_pattern, "**/", pattern)
```

Comment:
It looks like `pattern` is not defined until afterwards, or am I missing something?

Reply:
No sorry my bad, the `pattern` variable seems defined in the outer function before the function call. That being said, it might be clearer if `pattern` was an explicit argument to the inner function (which sort of echoes @NicolasGensollen's comment).

Reply:
This will work but it would be cleaner to pass `pattern` as a proper arg (also it won't work if we decide to extract it out of the outer function as suggested in my previous comment).

Reply:
If you keep it inside, `find_sub_ses_pattern` becomes a closure which *inherits* the context of the caller (including `pattern`). If you move it out, it becomes a standalone function and needs a proper parameter for passing the value.

It's just to say @14thibea that you did nothing wrong, both usages work. For testing purposes however, I second @NicolasGensollen and @omar-rifai recommendations to extract the code to a separate function instead of using a closure.